File: p2/solutions/Lambda2.java
class AccountCompensation{
  public void withdrawUndo(double amount){
    acc.deposit(amount);
  }
  public void depositUndo(double amount){
    acc.withdraw(amount);
  }
}
class AccountService
{
  Stack<Lamda> stack = new Stack<Lamda>();
  Account acc = new Account();
  AccountCompensation compensation = new AccountCompensation();
  public void withdraw(double amount){
    acc.withdraw(amount);
    stack.push(()->compensation.withdrawUndo(amount));
  }
  public void deposit(double amount){
    acc.deposit(amount);
    stack.push(()->compensation.depositUndo(amount));
  }
  public void undo(){
    Lamda fun = stack.pop();
    fun();
  }
}

File: p2/solutions/Lambda2.py
class AccountService:
    def __init__(self):
        self.stack = Stack()
        self.acc = Account()

    def withdraw(self, amount):
        self.acc.withdraw(amount)
        self.stack.push(lambda : withdrawUndo(amount))

    def deposit(self, amount):
        self.acc.deposit(amount)
        self.stack.push(lambda : depositUndo(amount))

    def undo(self):
        op = self.stack.pop()
        op()

    def withdrawUndo(self, amount):
        self.acc.deposit(amount)

    def depositUndo(self, amount):
        self.acc.withdraw(amount)
 
